DynamicActiveQuery.php

```diff
@@ -27,6 +28,44 @@ class DynamicActiveQuery extends ActiveQuery
      */
     private $dynamicColumn;
 
+    private $wrapped = false;
+
+    /**
+     * Convert index value to closurem that will get decoded dynamic attribute, in case if indexing attribute is dynamic
+     * @param callable|string $column
+     * @return $this
+     */
+    public function indexBy($column)
+    {
+        if ($this->asArray) {
+            /** @var \yii\db\ActiveRecord $modelClass */
+            $modelClass = $this->modelClass;
+            $this->indexBy = function ($row) use ($column, $modelClass) {
+                if (isset($row[$column])) {
+                    return $row[$column];
+                } elseif (method_exists($modelClass, 'dynamicColumn')) {
```

Comment:
please don't use [else after return](https://github.com/yiisoft/yii2/blob/master/docs/internals/core-code-style.md#54-control-statements). it adds indent and makes the logic less clear.